aptos-move/block-executor/src/executor.rs

```diff
@@ -424,46 +393,32 @@ where
         // TODO: should not need to pass base view.
         base_view: &S,
         shared_counter: &AtomicU32,
-        role: CommitRole,
+        txn_fee_state: &Mutex<(FeeStatement, Vec<FeeStatement>)>,
     ) {
         // Make executor for each task. TODO: fast concurrent executor.
         let init_timer = VM_INIT_SECONDS.start_timer();
         let executor = E::init(*executor_arguments);
         drop(init_timer);
 
-        let committing = matches!(role, CommitRole::Coordinator(_));
-
         let _timer = WORK_WITH_TASK_SECONDS.start_timer();
         let mut scheduler_task = SchedulerTask::NoTask;
-        let mut worker_idx = 0;
 
-        let mut accumulated_fee_statement = FeeStatement::zero();
-        let mut txn_fee_statements = Vec::with_capacity(block.len());
         loop {
-            // Only one thread does try_commit to avoid contention.
-            match &role {
-                CommitRole::Coordinator(post_commit_txs) => {
-                    self.coordinator_commit_hook(
-                        self.maybe_block_gas_limit,
-                        scheduler,
-                        post_commit_txs,
-                        &mut worker_idx,
-                        &mut scheduler_task,
-                        last_input_output,
-                        &mut accumulated_fee_statement,
-                        &mut txn_fee_statements,
-                    );
-                },
-                CommitRole::Worker(rx) => {
-                    while let Ok(txn_idx) = rx.try_recv() {
-                        self.worker_commit_hook(
-                            txn_idx,
-                            versioned_cache,
-                            last_input_output,
-                            base_view,
-                        );
-                    }
-                },
+            if scheduler.should_coordinate_commits() {
```

Comment:
this would have to be while scheduler.should_coordinate_commits() {}

otherwise, the thread unlocks after someone armed (but couldn't get the lock), they both get stuck on dependencies while executing the next task right after (before coming back to the above loop), and if progress depends on another shard unblocking us, that shard may be waiting for the commit notification.
If this was a while, then one of them will for sure commit anything that can be committed (based on the latest arm)

Reply:
@zjma : this is re:resolving the cross-shard coordinating-related deadlock. The PR either solves it already, or should be really close, if you also want to have a look / confirm / join the discussion.

Reply:
Took a quick look. i think this solves the sharding issue. Will run some separate experiments to verify.

Reply:
Cool